web/src/resources/TagResource.ts

```diff
@@ -266,62 +279,112 @@ export async function bulkDeleteTags(
-  error: Error;
+export class TagOperationError extends ResourceError {
   tag: string;
   constructor(message: string, tag: string, error: AxiosError) {
-    super(message);
+    super(message, tag, error);
     this.tag = tag;
     this.error = error;
-    Object.setPrototypeOf(this, TagDeleteError.prototype);
+    Object.setPrototypeOf(this, TagOperationError.prototype);
   }
 }
 
-export async function deleteTag(org: string, repo: string, tag: string) {
+export async function deleteTag(org: string, repo: string, tag: Tag) {
   try {
     const response: AxiosResponse = await axios.delete(
-      `/api/v1/repository/${org}/${repo}/tag/${tag}`,
+      `/api/v1/repository/${org}/${repo}/tag/${tag.name}`,
     );
     assertHttpCode(response.status, 204);
   } catch (err) {
-    throw new TagDeleteError(
+    throw new TagOperationError(
       'failed to delete tag',
-      `${org}/${repo}:${tag}`,
+      `${org}/${repo}:${tag.name}`,
       err,
     );
   }
 }
 
-export async function expireTag(org: string, repo: string, tag: string) {
+export async function expireTag(org: string, repo: string, tag: Tag) {
   try {
     const response: AxiosResponse = await axios.post(
-      `/api/v1/repository/${org}/${repo}/tag/${tag}/expire`,
+      `/api/v1/repository/${org}/${repo}/tag/${tag.name}/expire`,
       {
         include_submanifests: true,
         is_alive: true,
       },
     );
     assertHttpCode(response.status, 200);
   } catch (err) {
-    throw new TagDeleteError(
+    throw new TagOperationError(
       'failed to expire tag',
+      `${org}/${repo}:${tag.name}`,
+      err,
+    );
+  }
+}
+
+export async function setTagMutability(
+  org: string,
+  repo: string,
+  tag: string,
+  immutability = false,
+) {
+  try {
+    const response: AxiosResponse = await axios.put(
+      `/api/v1/repository/${org}/${repo}/tag/${tag}`,
+      {
+        immutable: immutability,
+      },
+    );
+    assertHttpCode(response.status, 201);
+  } catch (err) {
+    throw new TagOperationError(
+      'failed to set tag to ' + (immutability ? 'immutable' : 'mutable'),
       `${org}/${repo}:${tag}`,
       err,
     );
   }
 }
 
+export async function setTagsMutability(
+  tags: TagLocation[],
+  immutability = false,
+) {
+  const responses = await Promise.allSettled(
+    tags.map((tag) =>
+      setTagMutability(tag.org, tag.repo, tag.tag, immutability),
+    ),
+  );
+
+  // Filter failed responses
+  const errResponses = responses.filter(
+    (r) => r.status == 'rejected',
+  ) as PromiseRejectedResult[];
+
+  // If errors collect and throw
+  if (errResponses.length > 0) {
+    const bulkDeleteError = new BulkOperationError<TagOperationError>(
+      'error setting tags to ' + (immutability ? 'immutable' : 'mutable'),
+    );
+    for (const response of errResponses) {
+      const reason = response.reason as TagOperationError;
+      bulkDeleteError.addError(reason.tag, reason);
+    }
+    throw bulkDeleteError;
+  }
+}
```

Comment:
You could probably use [throwIfError](https://github.com/quay/quay/blob/master/web/src/resources/ErrorHandling.ts#L31) here

Reply:
Cool, TIL. Updated.,